test/stress/s3benchmark.cc

```diff
@@ -0,0 +1,225 @@
+    if (duplicate_chance < duplicate_file_ratio_ and filenames_.size() > 0) {
+      int duplicate_index = rand() % filenames_.size();
+      filenames_.push_back(filenames_[duplicate_index]);
+    } else {
+      ++unique_files_;
+      filenames_.push_back(GenerateFile());
+    }
+  }
+}
+
+std::string S3TestScenario::GenerateFile()
+{
+  std::string filename;
+  do {
+    filename = GenerateRandomFilename();
+  } while(FileExists(tmp_path_ + "/" + filename));
+  std::string file_path = tmp_path_ + "/" + filename;
+
+  int file_size = 0;
+  while(file_size <= 0) file_size = avg_file_size_ + int(RandomNormal()*avg_file_size_*relative_variance_);
+
+  int chunk_size = 1024;
+  int bytes_written = 0;
+  char *buffer = reinterpret_cast<char *>(smalloc(sizeof(char)*(chunk_size + 1)));
+  buffer[chunk_size] = 0;
+  std::ofstream random_file;
+
+  random_file.open(file_path);
+  assert(!random_file.fail());
+
+  while(bytes_written < file_size) {
+    for(int i = 0; i < chunk_size; ++i) {
+      buffer[i] = rand() % (CHAR_MAX - CHAR_MIN) + CHAR_MIN;
+    }
+    int num_bytes_to_write = (chunk_size <= file_size - bytes_written) ? chunk_size : (file_size - bytes_written);
+    random_file.write(buffer, num_bytes_to_write);
+    bytes_written += num_bytes_to_write;
+  }
+  free(buffer);
+  random_file.close();
+  return filename;
+}
+
+std::string S3TestScenario::GenerateRandomFilename()
+{
+  int len = 10;
+  static const char alphanum[] =
+        "0123456789"
+        "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
+        "abcdefghijklmnopqrstuvwxyz";
+
+  std::string result(len, 0);
+  for(int i = 0; i < len; ++i) {
+    result[i] = alphanum[rand() % (sizeof(alphanum) - 1)];
+  }
+  return result;
+}
+
+void S3TestScenario::UploadFile(const std::string &filename)
+{
+  std::string file_path = tmp_path_ + "/" + filename;
+  uploader_->Upload(file_path, filename);
+}
+
+void S3TestScenario::Cleanup()
+{
+  assert(RemoveTree(tmp_path_));
+}
+
+// Gutted Box-Muller transform
+float S3TestScenario::RandomNormal() {
+  float z, u1, u2;
+  float pi = atan(1) * 4;
+  u1 = rand() * (1.0 / RAND_MAX);
+  u2 = rand() * (1.0 / RAND_MAX);
+  z = sqrt(-2.0 * log(u1)) * cos(2 * pi * u2);
+  return z;
+}
+
+
+int main(int argc, char *argv[])
+{
+  string s3_config, tmp_path = "/tmp/s3benchmark";
+  int num_files = 1000, file_size = 4096;
+  float duplicate_ratio = 0;
+
+  int c;
+  while((c = getopt(argc, argv, "c:n:s:t:d:")) != -1) {
+    switch(c) {
+      case 'c':
+        s3_config = string(optarg);
+        break;
+      case 'n':
+        num_files = atoi(optarg);
+        break;
+      case 's':
+        file_size = atoi(optarg);
+        break;
+      case 't':
+        tmp_path = string(optarg);
+      case 'd':
+        duplicate_ratio = atof(optarg);
+    }
+  }
+  srand(time(NULL));
+  S3TestScenario *scenario = new S3TestScenario(
+    s3_config,
+    tmp_path,
+    num_files,
+    file_size,
+    duplicate_ratio);
+
+  scenario->Run();
```

Comment:
That perhaps should rather be `return scenario->Run()` to accommodate for possible failures